src/Razor/Microsoft.AspNetCore.Razor.Language/src/Syntax/InternalSyntax/SyntaxTokenCache.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Razor.Language.Syntax.InternalSyntax
+{
+    // Simplified version of Roslyn's SyntaxNodeCache
+    internal static class SyntaxTokenCache
+    {
+        private const int CacheSizeBits = 16;
+        private const int CacheSize = 1 << CacheSizeBits;
+        private const int CacheMask = CacheSize - 1;
+        private static readonly Entry[] s_cache = new Entry[CacheSize];
+
+        private struct Entry
+        {
+            public int Hash { get; }
+            public SyntaxToken Token { get; }
+
+            internal Entry(int hash, SyntaxToken token)
+            {
+                Hash = hash;
+                Token = token;
+            }
+        }
+
+        public static bool CanBeCached(SyntaxKind kind, params RazorDiagnostic[] diagnostics)
+        {
+            if (diagnostics.Length == 0)
+            {
+                switch (kind)
+                {
+                    case SyntaxKind.CharacterLiteral:
+                    case SyntaxKind.Dot:
+                    case SyntaxKind.Identifier:
+                    case SyntaxKind.IntegerLiteral:
+                    case SyntaxKind.Keyword:
+                    case SyntaxKind.NewLine:
+                    case SyntaxKind.RazorCommentStar:
+                    case SyntaxKind.RazorCommentTransition:
+                    case SyntaxKind.StringLiteral:
+                    case SyntaxKind.Transition:
+                    case SyntaxKind.Whitespace:
+                        return true;
+                }
+            }
+
+            return false;
+        }
+
+        public static SyntaxToken GetCachedToken(SyntaxKind kind, string content)
+        {
+            var hash = (kind, content).GetHashCode();
+
+            // Allow the upper 16 bits to contribute to the index
+            var indexableHash = hash ^ (hash >> 16);
```

Comment:
Why this?

Reply:
And is this unrelated to `CacheSizeBits` which is also 16?

Reply:
The hashing calculation gives a 32 bit result. This allows me to let all 32 bits contribute to the calculated index. It's not really related to CacheSizeBits (other than the fact that CacheSizeBits is less that 32 and indicates that we need to do this to use the additional hashing data)

---
In reply to: [437672778](https://github.com/dotnet/aspnetcore/pull/22686#discussion_r437672778) [](ancestors = 437672778)